File: src/api/data/repository.ts
import { getDb } from './db';
import type { EtfBrief, SectorGroup, HomeResponse, HistoryDay, HistoryEtf } from '../types';

// ---- DB 行类型 ----

interface DbEtfScoreRow {
  etf_code: string;
  etf_name: string;
  composite_score: number;
  rank: number;
  sector: string;
  tags: string;
  market_tag: string;
  calc_date: string;
  created_at: string;
}

interface DbHistoricalRow {
  record_date: string;
  etf_code: string;
  etf_name: string;
  signal_3d_return: number | null;
  current_return: number | null;
}

// ---- 辅助 ----

function parseTags(tagsJson: string): string[] {
  try {
    return JSON.parse(tagsJson);
  } catch {
    return [];
  }
}

function formatReturn(value: number | null): string | null {
  if (value === null || value === undefined) return null;
  const prefix = value >= 0 ? '+' : '';
  return `${prefix}${value.toFixed(2)}%`;
}

// ---- 数据查询 ----

export function getTop5Scores(calcDate: string): DbEtfScoreRow[] {
  const db = getDb();
  return db
    .prepare('SELECT etf_code, etf_name, composite_score, rank, sector, tags, market_tag, calc_date, created_at FROM etf_scores WHERE calc_date = ? ORDER BY rank ASC LIMIT 5')
    .all(calcDate) as DbEtfScoreRow[];
}

export function getEtfScores(calcDate: string): DbEtfScoreRow[] {
  const db = getDb();
  return db
    .prepare('SELECT etf_code, etf_name, composite_score, rank, sector, tags, market_tag, calc_date, created_at FROM etf_scores WHERE calc_date = ? ORDER BY rank ASC')
    .all(calcDate) as DbEtfScoreRow[];
}

export function getLatestCalcDate(): string | undefined {
  const db = getDb();
  const row = db.prepare('SELECT calc_date FROM etf_scores ORDER BY calc_date DESC LIMIT 1').get() as { calc_date: string } | undefined;
  return row?.calc_date;
}

export function getHistoricalRecords(startDate?: string, endDate?: string): DbHistoricalRow[] {
  const db = getDb();
  let query = 'SELECT record_date, etf_code, etf_name, signal_3d_return, current_return FROM historical_records WHERE 1=1';
  const params: string[] = [];
  if (startDate) {
    query += ' AND record_date >= ?';
    params.push(startDate);
  }
  if (endDate) {
    query += ' AND record_date <= ?';
    params.push(endDate);
  }
  query += ' ORDER BY record_date DESC, rank ASC';
  return db.prepare(query).all(...params) as DbHistoricalRow[];
}

// ---- 业务对象构建 ----

function rowToEtfBrief(row: DbEtfScoreRow): EtfBrief {
  return {
    code: row.etf_code,
    name: row.etf_name,
    score: Math.round(row.composite_score * 100) / 100,
    rank: row.rank,
    sector: row.sector || '',
    tags: parseTags(row.tags),
  };
}

export function buildTop5Result(calcDate: string): HomeResponse | null {
  const rows = getTop5Scores(calcDate);
  if (rows.length === 0) return null;

  const topEtfs = rows.map(rowToEtfBrief);

  // 赛道分布
  const sectorMap = new Map<string, EtfBrief[]>();
  for (const etf of topEtfs) {
    const sec = etf.sector || '其他';
    if (!sectorMap.has(sec)) sectorMap.set(sec, []);
    sectorMap.get(sec)!.push(etf);
  }
  const sectors: SectorGroup[] = Array.from(sectorMap.entries()).map(([name, etfs]) => ({ name, etfs }));

  return {
    top5_etfs: topEtfs,
    sectors,
    signal_date: calcDate,
    market_tag: rows[0]?.market_tag || '数据未就绪',
    update_time: rows[0]?.created_at || new Date().toISOString(),
  };
}

export function buildHistory(): HistoryDay[] {
  const records = getHistoricalRecords();
  const dayMap = new Map<string, HistoryEtf[]>();

  for (const r of records) {
    if (!dayMap.has(r.record_date)) dayMap.set(r.record_date, []);
    dayMap.get(r.record_date)!.push({
      code: r.etf_code,
      name: r.etf_name,
      signal_3d_return: formatReturn(r.signal_3d_return),
      current_return: formatReturn(r.current_return),
    });
  }

  const result: HistoryDay[] = [];
  for (const [date, etfs] of dayMap.entries()) {
    result.push({ date, etfs });
  }
  result.sort((a, b) => b.date.localeCompare(a.date));
  return result;
}

export function buildHistoryForDate(date: string): HistoryDay | null {
  const records = getHistoricalRecords(date, date);
  if (records.length === 0) return null;

  return {
    date,
    etfs: records.map((r) => ({
      code: r.etf_code,
      name: r.etf_name,
      signal_3d_return: formatReturn(r.signal_3d_return),
      current_return: formatReturn(r.current_return),
    })),
  };
}

File: src/morning_report/data/repository.py
"""数据访问层 — SQLite 存储，Repository/DAO 模式。

Status: [default]
"""

import json
import sqlite3
from contextlib import contextmanager
from datetime import date, datetime
from pathlib import Path
from typing import Optional

from ..config import DB_PATH
from ..models import ETFScore, HistoricalRecord


def get_db_path() -> Path:
    """获取数据库文件路径。"""
    import os
    # 尝试从环境变量获取 git root，否则使用当前目录
    git_root = os.environ.get("GIT_ROOT", "")
    if git_root:
        base = Path(git_root)
    else:
        base = Path.cwd()
    return base / DB_PATH


@contextmanager
def get_connection():
    """获取数据库连接（自动提交/关闭）。"""
    db_path = get_db_path()
    conn = sqlite3.connect(str(db_path))
    conn.row_factory = sqlite3.Row
    try:
        yield conn
        conn.commit()
    except Exception:
        conn.rollback()
        raise
    finally:
        conn.close()


def init_db():
    """初始化数据库表结构 [default]."""
    with get_connection() as conn:
        conn.executescript("""
            CREATE TABLE IF NOT EXISTS etf_scores (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                calc_date TEXT NOT NULL,
                etf_code TEXT NOT NULL,
                etf_name TEXT NOT NULL,
                heat_rank_score REAL NOT NULL,
                momentum_rank_score REAL NOT NULL,
                liquidity_rank_score REAL NOT NULL,
                low_vol_rank_score REAL NOT NULL,
                composite_score REAL NOT NULL,
                rank INTEGER NOT NULL,
                sector TEXT DEFAULT '',
                tags TEXT DEFAULT '[]',
                market_tag TEXT DEFAULT '',
                status TEXT DEFAULT 'success',
                created_at TEXT NOT NULL DEFAULT (datetime('now')),
                UNIQUE(calc_date, etf_code)
            );

            CREATE INDEX IF NOT EXISTS idx_etf_scores_date
                ON etf_scores(calc_date);

            CREATE INDEX IF NOT EXISTS idx_etf_scores_rank
                ON etf_scores(calc_date, rank);

            CREATE TABLE IF NOT EXISTS historical_records (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                record_date TEXT NOT NULL,
                etf_code TEXT NOT NULL,
                etf_name TEXT NOT NULL,
                composite_score REAL NOT NULL,
                rank INTEGER NOT NULL,
                signal_3d_return REAL,
                current_return REAL,
                created_at TEXT NOT NULL DEFAULT (datetime('now')),
                UNIQUE(record_date, etf_code)
            );

            CREATE INDEX IF NOT EXISTS idx_historical_date
                ON historical_records(record_date);
        """)


def save_etf_scores(scores: list[ETFScore]):
    """保存 ETF 评分结果（幂等：相同 calc_date + etf_code 覆盖）."""
    with get_connection() as conn:
        conn.executemany("""
            INSERT OR REPLACE INTO etf_scores
                (calc_date, etf_code, etf_name,
                 heat_rank_score, momentum_rank_score,
                 liquidity_rank_score, low_vol_rank_score,
                 composite_score, rank, sector, tags,
                 market_tag, status, created_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, datetime('now'))
        """, [
            (
                s.calc_date.isoformat(), s.etf_code, s.etf_name,
                s.heat_rank_score, s.momentum_rank_score,
                s.liquidity_rank_score, s.low_vol_rank_score,
                s.composite_score, s.rank,
                s.sector, json.dumps(s.tags, ensure_ascii=False),
                s.market_tag.value if s.market_tag else "",
                s.status.value,
            )
            for s in scores
        ])


def get_etf_scores(calc_date: date) -> list[dict]:
    """查询指定日期的 ETF 评分（按 rank ASC）."""
    with get_connection() as conn:
        rows = conn.execute(
            "SELECT * FROM etf_scores WHERE calc_date = ? ORDER BY rank ASC",
            (calc_date.isoformat(),)
        ).fetchall()
        result = []
        for r in rows:
            d = dict(r)
            d["tags"] = json.loads(d.get("tags", "[]"))
            result.append(d)
        return result


def get_top5_scores(calc_date: date) -> list[dict]:
    """查询指定日期的 Top5 ETF 评分."""
    with get_connection() as conn:
        rows = conn.execute(
            "SELECT * FROM etf_scores WHERE calc_date = ? ORDER BY rank ASC LIMIT 5",
            (calc_date.isoformat(),)
        ).fetchall()
        result = []
        for r in rows:
            d = dict(r)
            d["tags"] = json.loads(d.get("tags", "[]"))
            result.append(d)
        return result


def save_historical_records(records: list[HistoricalRecord]):
    """保存历史表现记录."""
    with get_connection() as conn:
        conn.executemany("""
            INSERT OR REPLACE INTO historical_records
                (record_date, etf_code, etf_name, composite_score,
                 rank, signal_3d_return, current_return, created_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, datetime('now'))
        """, [
            (
                r.date.isoformat(), r.etf_code, r.etf_name,
                r.composite_score, r.rank,
                r.signal_3d_return, r.current_return,
            )
            for r in records
        ])


def get_historical_records(
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
) -> list[dict]:
    """查询历史表现记录."""
    with get_connection() as conn:
        query = "SELECT * FROM historical_records WHERE 1=1"
        params = []
        if start_date:
            query += " AND record_date >= ?"
            params.append(start_date.isoformat())
        if end_date:
            query += " AND record_date <= ?"
            params.append(end_date.isoformat())
        query += " ORDER BY record_date DESC, rank ASC"
        rows = conn.execute(query, params).fetchall()
        return [dict(r) for r in rows]
